tools/clang/unittests/HLSL/CompilerTest.cpp

```diff
@@ -2589,6 +2603,80 @@ TEST_F(CompilerTest, CompileWhenVdThenProducesDxilContainer) {
+                   &pSource);
+  pInclude = new TestIncludeHandler(m_dllSupport);
+  pInclude->CallResults.emplace_back(includedData, includedSize, 0);
+
+  const WCHAR *pArgs[] = {L"-encoding",
+                          encoding};
+  VERIFY_SUCCEEDED(pCompiler->Compile(pSource, L"source.hlsl", L"main",
+                                      L"ps_6_0", pArgs, _countof(pArgs),
+                                      nullptr, 0, pInclude, &pResult));
+  HRESULT status;
+  VERIFY_SUCCEEDED(pResult->GetStatus(&status));
+
+  if (succ) {
+    VERIFY_SUCCEEDED(status);
+  } else {
+    CComPtr<IDxcBlobEncoding> pErrors;
+    VERIFY_ARE_EQUAL(status, DXC_E_STRING_ENCODING_FAILED);
+    VERIFY_SUCCEEDED(pResult->GetErrorBuffer(&pErrors));
+    std::string textUtf8 = BlobToUtf8(pErrors);
+    LPCSTR pErrorMsg = "Error in encoding argument specified";
+    CheckOperationResultMsgs(pResult, &pErrorMsg, 1, false,
+                             false);
+  }
+}
+
+TEST_F(CompilerTest, CompileWithEncodeFlagTestSource) {
+  // Testing for utf8 encoded files
+  std::string sourceUtf8 = "#include \"include.hlsl\"\r\n"
+                             "float4 main() : SV_Target { return 0; }";
+  std::string includeUtf8 = "// Comment\n";
+  std::string utf8BOM = "\xEF" "\xBB" "\xBF"; // UTF-8 BOM
+  std::string includeUtf8BOM = utf8BOM + includeUtf8;
+
+  TestEncodingImpl(sourceUtf8.data(), sourceUtf8.size(), DXC_CP_UTF8,
+                   includeUtf8BOM.data(), includeUtf8BOM.size(),
+                   true, L"utf16");
+
+  TestEncodingImpl(sourceUtf8.data(), sourceUtf8.size(), DXC_CP_UTF8,
+                   includeUtf8.data(), includeUtf8.size(), false, L"utf16");
+
+  TestEncodingImpl(sourceUtf8.data(), sourceUtf8.size(), DXC_CP_UTF8,
+                   includeUtf8.data(), includeUtf8.size(), true, L"utf8");
+
+  // Testing for utf16 encoded files
+  std::wstring sourceUtf16 = L"#include \"include.hlsl\"\r\n"
+                             L"float4 main() : SV_Target { return 0; }";
+  std::wstring includeUtf16 = L"// Comments\n";
+  std::wstring utf16BOM = L"\xFEFF"; // UTF-16 LE BOM
+  std::wstring includeUtf16BOM = utf16BOM + includeUtf16;
+
+  TestEncodingImpl(sourceUtf16.data(), sourceUtf16.size() * sizeof(L'A'),
+                   DXC_CP_UTF16,
+                   includeUtf16BOM.data(), includeUtf16BOM.size() * sizeof(L'A'), true, L"utf8");
+
+   //TO-DO: not hitting a custom error here, need to add a test that checks if a file specified
+   //as utf8 is actually utf encoded earlier in compilation
+  //TestEncodingImpl(sourceUtf16.data(), sourceUtf16.size() * sizeof(L'A'),
+  //                 DXC_CP_UTF16,
+  //                 includeUtf16.data(), includeUtf16.size() * sizeof(L'A'), false, L"utf8");
+
+  TestEncodingImpl(sourceUtf16.data(), sourceUtf16.size() * sizeof(L'A'),
+                   DXC_CP_UTF16,
+                   includeUtf16.data(), includeUtf16.size() * sizeof(L'A'),
+                   true, L"utf16");
```

Comment:
One thing that might be missing is the mixed case where the input buffer is one type, but the default and included file is a different type.  That seems like it could be interesting to test in case there are any assumptions between the main source file input format and the included format.